app/src/main/java/com/infomaniak/swisstransfer/upload/TransferUploader.kt

```diff
@@ -199,101 +199,87 @@ class TransferUploader(
-                        allChunksButLastUploadedSignal.complete()
-                    }
-                    continue
-                }
-
-                requestSemaphore.acquire()
-
-                val isLastChunk = chunkIndex == lastChunkIndex
-                val dataByteArray = getReusableByteArray(byteArrayPool, inputStream, chunkSize, isLastChunk)
-
-                val isEndOfFile = inputStream.read(dataByteArray, 0, dataByteArray.size) == -1
-                if (isEndOfFile) {
-                    SentryLog.w(TAG, "endOfFile for chunk #$chunkIndex of $fileUUID")
-                    continue
-                }
-                launch {
-                    try {
-
-                        // Wait for all the other jobs to complete
-                        if (totalChunks > 1 && isLastChunk) {
-                            allChunksButLastUploadedSignal.join()
-                        }
-                        metadata.chunksUploadStatus[chunkIndex] = StartedOrComplete
-                        SentryLog.d(TAG, "Uploading chunk #$chunkIndex of $fileUUID")
-                        startUploadChunk(
-                            fileUUID = fileUUID,
-                            chunkIndex = chunkIndex,
-                            isLastChunk = isLastChunk,
-                            data = dataByteArray
-                        )
-                        SentryLog.d(TAG, "Uploaded chunk #$chunkIndex of $fileUUID")
-                        metadata.chunksUploadStatus[chunkIndex] = DefinitelyComplete
-                        // Allow the last chunk to start being uploaded
-                        if (totalChunks > 1 && completedChunks.incrementAndFetch() == lastChunkIndex) {
-                            allChunksButLastUploadedSignal.complete()
-                        }
-                    } finally {
-                        byteArrayPool.offer(dataByteArray)
-                        requestSemaphore.release()
-                    }
+            for (chunkIndex in 0 until lastChunkIndex) launch {
+                requestSemaphore.withPermit {
+                    uploadChunkIfNeeded(metadata = metadata, chunkIndex = chunkIndex, isLastChunk = false)
                 }
             }
         }
-
-        byteArrayPool.clear()
+        uploadChunkIfNeeded(metadata = metadata, chunkIndex = lastChunkIndex, isLastChunk = true)
     }
 
-    private fun getReusableByteArray(
-        byteArrayPool: ArrayBlockingQueue<ByteArray>,
-        inputStream: BufferedInputStream,
-        chunkSize: Long,
-        isLastChunk: Boolean,
-    ): ByteArray {
-        val currentChunkSize = if (isLastChunk) inputStream.available() else chunkSize.toInt()
-        val reusableByteArray = if (isLastChunk) null else byteArrayPool.poll()
+    private suspend fun uploadChunkIfNeeded(metadata: FileToUploadMetaData, chunkIndex: Int, isLastChunk: Boolean) {
+        val uploadStatus = metadata.chunksUploadStatus[chunkIndex]
```

Comment:
this, `targetFileUri`and `totalFileSize` are used only once, we can remove them

Reply:
Not sold on the readability improvement, but inlined `targetFileUri`.
Left `totalFileSize` to not make the line below too long and difficult to read.